File: notification/notification_reader.py
#!/data/data/com.termux/files/usr/bin/python
import sys
import re

packages = {
    'com.tencent.mm': '微信',
    'com.tencent.tim': 'qq',
    'com.tencent.wework': '企业微信',
}


class WechatNotificationParser:
    MSG_TYPE_GROUP = 1
    MSG_TYPE_PERSONAL = 2
    MSG_NOTIFICATION = 3  # 非消息类型的微信通知

    msg_from = None  # 比如某人或某群
    msg_type = None  # 个人消息或群消息
    msg_publisher = None  # 发消息的人
    msg_num = 1  # 当前未读消息数量
    msg_content = None  # 消息内容

    def __init__(self, notification: str):
        splits = notification.split(',')
        raw_content = re.search('.+(?=,%evtprm4,%evtprm5,%evtprm6)', ','.join(splits[2:])).group()
        self.msg_from = splits[1]
        if ': ' in raw_content:
            content_splits = raw_content.split(': ')
            if re.match('^\\[\\d+条].+$', content_splits[0]):
                self.msg_publisher = re.search('(\\[\\d+条])?(.+)', content_splits[0]).groups()[1]
                self.msg_num = re.search('(?<=\\[)\\d+(?=条])', content_splits[0]).group()
            else:
                self.msg_publisher = content_splits[0]
                self.msg_num = 1
            self.msg_content = content_splits[1]
            self.msg_type = WechatNotificationParser.MSG_TYPE_GROUP if self.msg_from != self.msg_publisher \
                else WechatNotificationParser.MSG_TYPE_PERSONAL
        else:
            self.msg_type = WechatNotificationParser.MSG_TYPE_PERSONAL
            self.msg_publisher = splits[1]
            self.msg_num = 1
            self.msg_content = raw_content
        if re.search(',%evtprm\\d,msg,false', ','.join(splits[3:])) is None:
            self.msg_type = WechatNotificationParser.MSG_NOTIFICATION

    @staticmethod
    def find(predicate, iterable) -> object:
        for i in iterable:
            if predicate(i):
                return i
        return None

    @staticmethod
    def action_message(msg: str) -> str:
        types = ['图片', '动画表情', '链接', '语音', '文件', '视频']
        num_word = ['一张', '一张', '一条', '一段', '一个', '一段']
        if re.match('\\[.{1,4}]', msg):  # 没有文字内容
            _type = WechatNotificationParser.find(lambda x: x in msg, types)
            if _type is not None:
                return f'发布了{num_word[types.index(_type)]}{_type}'
            else:
                return f'发布了一个表情，{msg}'
        else:
            return '说，' + msg


def read(notification: str) -> str:
    notification_splits = notification.split(',')
    application = packages[notification_splits[0]]
    if application == '微信':
        results = WechatNotificationParser(notification)
        action_meg = WechatNotificationParser.action_message(results.msg_content)
        if results.msg_type == WechatNotificationParser.MSG_TYPE_GROUP:
            return f'{application}消息，{results.msg_publisher}在群{results.msg_from}中{action_meg}'
        elif results.msg_type == WechatNotificationParser.MSG_TYPE_PERSONAL:
            return f'{application}消息，{results.msg_publisher}{action_meg}'
        else:
            return f'{application}通知，{results.msg_from},{results.msg_content}'
    else:
        if application is None:
            application = '应用' + notification_splits[0]
        return f'{application}消息，{notification_splits[1]}' + ','.join(notification_splits[2:])


if __name__ == '__main__':
    notification = ' '.join(sys.argv[1:])
    print(read(notification))

File: notification/notification_reader.js
const MSG_TYPE_GROUP = 1;
const MSG_TYPE_PERSONAL = 2;
const MSG_NOTIFICATION = 3;

function parse(notification) {
    let splits = notification.split(",");
    let rawContent = splits.slice(2).join(",").match(/.+(?=,%evtprm4,%evtprm5,%evtprm6)/)[0];
    let msgFrom = splits[1];
    var msgType, msgPublisher, msgNum, msgContent;

    if (rawContent.includes(": ")) {
        let contentSplits = rawContent.split(": ");

        if (/^\[\d+条].+$/.test(contentSplits[0])) {
            msgPublisher = contentSplits[0].match(/(\[\d+条])?(.+)/)[2];
            msgNum = contentSplits[0].match(/(?<=\[)\d+(?=条])/)[0];
        } else {
            msgPublisher = contentSplits[0];
            msgNum = 1;
        }

        msgContent = contentSplits[1];
        msgType = msgFrom !== msgPublisher ? MSG_TYPE_GROUP : MSG_TYPE_PERSONAL;
    } else {
        msgType = MSG_TYPE_PERSONAL;
        msgPublisher = splits[1];
        msgNum = 1;
        msgContent = rawContent;
    }

    if (!/,%evtprm\d,msg,false/.test(splits.slice(3).join(","))) {
        msgType = MSG_NOTIFICATION;
    }

    return {
        from: msgFrom,
        type: msgType,
        publisher: msgPublisher,
        num: msgNum,
        content: msgContent
    };
}

function actionMessage(msg) {
    let types = ["图片", "动画表情", "链接", "语音", "文件", "视频"];
    let numWord = ["一张", "一张", "一条", "一段", "一个", "一段"];

    if (/.*@.+ .*/.test(msg)) {
        who = msg.match(/(?<=@).+(?= )/)[0];
        msg = msg.replace(/@.+? /, "");
        return `@了${who}并且说，${msg}`;
    } 
    if (/\[.{1,4}]/.test(msg)) {
        let type = types.find(v => msg.includes(v));
        if (type !== undefined) {
            return `发布了${numWord[types.indexOf(type)]}${type}`;
        } else {
            return `发布了一个表情，${msg}`;
        }
    }   
    return "说，" + msg;
}

function read(notification) {
    const packages = {
        "com.tencent.mm": "微信",
        "com.tencent.tim": "qq",
        "com.tencent.wework": "企业微信"
    };

    let notificationSplits = notification.split(",");
    let application = packages[notificationSplits[0]];

    if (application === "微信") {
        let results = parse(notification);
        let actionMeg = actionMessage(results.content);

        switch(results.type) {
            case MSG_TYPE_GROUP:
                return `${application}消息，${results.publisher}在群${results.from}中${actionMeg}`;
            case MSG_TYPE_PERSONAL:
                return `${application}消息，${results.publisher}${actionMeg}`;
            default:
                if (results.content.includes("语音通话中")) {
                    return `${application}提醒，和${results.from}语音通话中`;
                } else {
                    return `${application}通知，${results.from},${results.content}`;
                }
        }
    } else {
        if (application === null) {
            application = "应用" + notificationSplits[0];
        }
        return `${application}消息，${notificationSplits[1]}` + notificationSplits.slice(2).join(",");
    }
}

 var output = read(rawmsg);; // 输出变量%output, task中局部变量名应为%rawmsg

//console.log(read("com.tencent.mm,GIANT心電途自行車俱樂部,乐益～老湿人: @凤仙 明早7：30分凤凰弯等你[呲牙]？,%evtprm4,%evtprm5,%evtprm6,msg,false"));

// function main() {
//     let text = process.argv[2];
//     console.log(`raw: ${text}`);
//     // let text = "com.tencent.wework,打卡,下班打卡提醒,%evtprm4,%evtprm5,%evtprm6,msg,true";
//     console.log(`msg: ${read(text)}`);
//     console.log()
// }

// main();
